torch/ao/quantization/fx/prepare.py

```diff
@@ -1129,38 +1200,136 @@ def insert_observers_for_model(
     #
     cache_for_no_tensor_check: Dict[Node, bool] = {}
 
-    inputs_seen_counter = 0
-    outputs_seen_counter = 0
-
     # first, populate the dtype map based only on qconfig and qhandler
     # this assumes:
     # graph inputs are fp32 by default, and int8 where overriden
     # other nodes output dtype is specified by the qconfig
     named_modules = dict(model.named_modules(remove_duplicate=False))
+
+    input_quantized_idxs: List[int] = prepare_custom_config.input_quantized_indexes
+    output_quantized_idxs: List[int] = prepare_custom_config.output_quantized_indexes
+    processed_nodes: Set[Node] = set()
+    # initalize target_dtype_info
+    DEFAULT_QCONFIG_FOR_TARGET_DTYPE_INFO = (
+        QConfig(
+            activation=PlaceholderObserver.with_args(dtype=torch.float32),
+            weight=PlaceholderObserver.with_args(dtype=torch.float32)
+        )
+    )
+    for node in model.graph.nodes:
+        node.meta["target_dtype_info"] = {
+            "input_act_obs_or_fq_ctr": DEFAULT_QCONFIG_FOR_TARGET_DTYPE_INFO.activation,
+            "output_act_obs_or_fq_ctr": DEFAULT_QCONFIG_FOR_TARGET_DTYPE_INFO.activation
+        }
+
+    inputs_seen_counter = 0
+    outputs_seen_counter = 0
+    placeholder_node_to_input_index: Dict[Node, int] = {}
+    # TODO: we probably don't need this counter since each graph will only have
+    # one output node?
+    output_node_to_output_index: Dict[Node, int] = {}
     for node in model.graph.nodes:
-        root_node, _, pattern, qhandler, qconfig = matches.get(
-            node.name, (None, None, None, None, None))
-        input_quantized_idxs: List[int] = prepare_custom_config.input_quantized_indexes
-        output_quantized_idxs: List[int] = prepare_custom_config.output_quantized_indexes
-        target_dtype_info: Dict[str, Optional[Tuple[Union[torch.dtype, type], bool]]] = \
-            _get_target_activation_dtype_for_node(
-                node, qconfig, inputs_seen_counter, outputs_seen_counter,
-                input_quantized_idxs, output_quantized_idxs, qhandler,
-                named_modules, cache_for_no_tensor_check)
-        node.meta["target_dtype_info"] = target_dtype_info
         if node.op == "placeholder":
+            placeholder_node_to_input_index[node] = inputs_seen_counter
             inputs_seen_counter += 1
         if node.op == "output":
+            output_node_to_output_index[node] = outputs_seen_counter
             outputs_seen_counter += 1
 
-    # Second, for nodes with known input dtypes, propagate them throughout the
+    # Step 1, set the observer or fake quantize module constructor for each node in the
+    # matched_node_pattern
```

Comment:
for my own understanding, should this be "each node in pattern", or "subgraph endpoints"?  I thought it was subgraph endpoints

Reply:
right now it is still each node in pattern, in the future we'll change this to "subgraph endpoints"